### PR TITLE
Fix query preview endpoint to be consistent with rest of query API

### PR DESCRIPTION
**Pull request by mike12345567**
## Description
As part of reviewing https://github.com/Budibase/budibase/pull/12772 with @melohagan - we decided that it would be beneficial if the query API could make use of subtypes, similar to how the table schema works. The query API is mostly capable of this, however the `preview` endpoint returned a flattened schema structure. The preview is the only part of the API which didn't really support returning more than just a string type, so the idea was to fix this.

The majority of this work involved cleaning up the `QueryViewer` and `RestQueryViewer` so that the preview response being a more complex structure (minimum `{ type: "string" }` for example) didn't cause issues.

## Addresses
- https://linear.app/budibase/issue/BUDI-7916/issue-with-how-the-schema-is-saved-between-rest-and-other-query-types

### CODE REVIEW

File: packages/builder/src/components/integration/RestQueryViewer.svelte [hunk 1]
```diff
@@ -386,6 +394,7 @@
 
   onMount(async () => {
     query = getSelectedQuery()
+    schema = query.schema
```
Comment (melohagan) [inline packages/builder/src/components/integration/RestQueryViewer.svelte:397]: On line 100 we have the following code:

```js
$: hasSchema =
    Object.keys(schema || {}).length !== 0 ||
    Object.keys(query?.schema || {}).length !== 0
```

because the schema is now set in the onMount, are we able to simplify that to:

```js
$: hasSchema = Object.keys(schema || {}).length !== 0
```

File: packages/server/src/api/controllers/query/index.ts [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
 import sdk from "../../../sdk"
 import { QueryEvent } from "../../../threads/definitions"
-import { ConfigType, Query, UserCtx, SessionCookie } from "@budibase/types"
+import {
+  ConfigType,
+  Query,
+  UserCtx,
+  SessionCookie,
+  QuerySchema,
+  FieldType,
```
Comment (melohagan) [inline packages/server/src/api/controllers/query/index.ts:17]: Just a sanity check - what is the difference between 'FieldType' and 'FieldTypes' ?

File: packages/server/src/api/controllers/query/index.ts [hunk 3]
```diff
@@ -163,38 +169,42 @@ export async function preview(ctx: UserCtx) {
     }
 
     const { rows, keys, info, extra } = (await Runner.run(inputs)) as any
```
Comment (melohagan) [inline packages/server/src/api/controllers/query/index.ts:171]: Can we replace the `as any` here?

File: packages/server/src/api/controllers/query/index.ts [hunk 4]
```diff
@@ -163,38 +169,42 @@ export async function preview(ctx: UserCtx) {
...
+    const makeQuerySchema = (type: FieldType, name: string): QuerySchema => ({
+      type,
+      name,
+    })
     if (rows?.length > 0) {
       for (let key of [...new Set(keys)] as string[]) {
         const field = rows[0][key]
         let type = typeof field,
-          fieldType = FieldTypes.STRING
+          fieldMetadata = makeQuerySchema(FieldType.STRING, key)
```
Comment (melohagan) [inline packages/server/src/api/controllers/query/index.ts:181]: NAB: Why use a function here? Why not:

```ts
let fieldMetadata: QuerySchema
...
fieldMetadata = {
  type: FieldType.STRING,
  name: key
}
```

File: packages/server/src/api/controllers/query/index.ts [hunk 5]
```diff
@@ -163,38 +169,42 @@ export async function preview(ctx: UserCtx) {
...
+        previewSchema[key] = fieldMetadata
       }
     }
     // if existing schema, update to include any previous schema keys
     if (existingSchema) {
-      for (let key of Object.keys(schemaFields)) {
+      for (let key of Object.keys(previewSchema)) {
         if (existingSchema[key]?.type) {
-          schemaFields[key] = existingSchema[key].type
+          previewSchema[key] = existingSchema[key].type
```
Comment (melohagan) [inline packages/server/src/api/controllers/query/index.ts:207]: This would flatten the `QuerySchema`. Do we not want:

```ts
previewSchema[key] = existingSchema[key]
```

Or is this purely for backwards compatibility with existing schemas?

File: packages/server/src/sdk/app/queries/queries.ts [hunk 6]
```diff
@@ -3,6 +3,27 @@ import { processStringSync } from "@budibase/string-templates"
 import { context } from "@budibase/backend-core"
 import { getQueryParams, isProdAppID } from "../../../db/utils"
 import { BaseQueryVerbs } from "../../../constants"
+import { Query, QuerySchema } from "@budibase/types"
+
+function updateSchema(query: Query) {
```
Comment (melohagan) [inline packages/server/src/sdk/app/queries/queries.ts:8]: NAB: should you explicitly provide the return type?

File: packages/server/src/sdk/app/queries/queries.ts [hunk 7]
```diff
@@ -3,6 +3,27 @@ import { processStringSync } from "@budibase/string-templates"
...
+      schema[key] = { type: query.schema[key] as string, name: key }
+    } else {
+      schema[key] = query.schema[key] as QuerySchema
+    }
+  }
+  query.schema = schema
+  return query
+}
+
+function updateSchemas(queries: Query[]) {
```
Comment (melohagan) [inline packages/server/src/sdk/app/queries/queries.ts:24]: NAB: should you explicitly provide the return type?

**Review by melohagan — APPROVED** [review]
LGTM! Really nice refactor and unit tests 👍 

I left a couple of small comments, but I did pull down the branch and appears to be working well!

File: yarn.lock
Comment (adrinr) [inline yarn.lock:1]: Why these changes? There is no changes on any package.json

File: packages/builder/src/components/integration/KeyValueBuilder.svelte [hunk 8]
```diff
@@ -35,6 +35,7 @@
   export let bindingDrawerLeft
   export let allowHelpers = true
   export let customButtonText = null
+  export let compare = (option, value) => option === value
```
[the remark this answers is not in the document]
Comment (mike12345567) [reply packages/builder/src/components/integration/KeyValueBuilder.svelte:38]: This is the default operation, basically wanted to define the default so that it can easily be overridden.

File: yarn.lock
Comment (mike12345567) [reply yarn.lock:1]: Unsure, the `yarn.lock` has been changing wildly lately - I'm not quite sure of the cause!

File: packages/server/src/api/controllers/query/index.ts (hunk 2 above)
Comment (mike12345567) [reply packages/server/src/api/controllers/query/index.ts:17]: `FieldTypes` is the old non-TS way that we did this, it hasn't been fully cleaned up yet - I'll raise a separate PR to finally be done with it.

If an enumeration is plural, not singular then it is an old pre-TS type, which needs to be gotten rid of. It doesn't really impact anything, but just for consistency sake working on getting rid of them.

File: packages/server/src/api/controllers/query/index.ts (hunk 4 above)
Comment (mike12345567) [reply packages/server/src/api/controllers/query/index.ts:181]: Just for clarity sake, reduces the chance of typos/extra props being added, Typescript does protect against this, but its just a rule I follow since we work in both JS/TS (with the frontend).

File: packages/server/src/api/controllers/query/index.ts (hunk 5 above)
Comment (mike12345567) [reply packages/server/src/api/controllers/query/index.ts:207]: you're right - sorry missed this!

File: packages/server/src/api/controllers/query/index.ts (hunk 4 above)
Comment (melohagan) [reply packages/server/src/api/controllers/query/index.ts:181]: Ah the extra props is a good point. Kind of like constructor overloading; create an object with the specified parameters only.

File: packages/builder/src/components/integration/KeyValueBuilder.svelte (hunk 8 above)
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/builder/src/components/integration/KeyValueBuilder.svelte:38]: I meant that is already the default down the component tree
